File: problem1-json-parser/results/JavaScript/A/main.js
'use strict';

// JSON parser implemented by hand (no built-in JSON.parse).
// Throws an Error on invalid input.

function parseJSON(text) {
  let i = 0;
  const n = text.length;

  function error(msg) {
    throw new Error(msg + ' at position ' + i);
  }

  function skipWhitespace() {
    while (i < n) {
      const c = text[i];
      if (c === ' ' || c === '\t' || c === '\n' || c === '\r') {
        i++;
      } else {
        break;
      }
    }
  }

  function parseValue() {
    skipWhitespace();
    if (i >= n) error('Unexpected end of input');
    const c = text[i];
    if (c === '{') return parseObject();
    if (c === '[') return parseArray();
    if (c === '"') return parseString();
    if (c === 't' || c === 'f') return parseBoolean();
    if (c === 'n') return parseNull();
    if (c === '-' || (c >= '0' && c <= '9')) return parseNumber();
    error('Unexpected token');
  }

  function parseObject() {
    i++; // consume '{'
    const obj = {};
    skipWhitespace();
    if (i < n && text[i] === '}') {
      i++;
      return obj;
    }
    while (true) {
      skipWhitespace();
      if (i >= n || text[i] !== '"') error('Expected string key');
      const key = parseString();
      skipWhitespace();
      if (i >= n || text[i] !== ':') error('Expected colon');
      i++; // consume ':'
      const value = parseValue();
      obj[key] = value;
      skipWhitespace();
      if (i >= n) error('Unexpected end of input in object');
      if (text[i] === ',') {
        i++;
        continue;
      }
      if (text[i] === '}') {
        i++;
        return obj;
      }
      error('Expected comma or closing brace');
    }
  }

  function parseArray() {
    i++; // consume '['
    const arr = [];
    skipWhitespace();
    if (i < n && text[i] === ']') {
      i++;
      return arr;
    }
    while (true) {
      const value = parseValue();
      arr.push(value);
      skipWhitespace();
      if (i >= n) error('Unexpected end of input in array');
      if (text[i] === ',') {
        i++;
        continue;
      }
      if (text[i] === ']') {
        i++;
        return arr;
      }
      error('Expected comma or closing bracket');
    }
  }

  function parseString() {
    i++; // consume opening '"'
    let result = '';
    while (i < n) {
      const c = text[i];
      if (c === '"') {
        i++;
        return result;
      }
      if (c === '\\') {
        i++;
        if (i >= n) error('Unexpected end of input in string');
        const esc = text[i];
        if (esc === '"') result += '"';
        else if (esc === '\\') result += '\\';
        else if (esc === 'n') result += '\n';
        else error('Invalid escape sequence');
        i++;
      } else {
        result += c;
        i++;
      }
    }
    error('Unterminated string');
  }

  function parseBoolean() {
    if (text.slice(i, i + 4) === 'true') {
      i += 4;
      return true;
    }
    if (text.slice(i, i + 5) === 'false') {
      i += 5;
      return false;
    }
    error('Invalid token');
  }

  function parseNull() {
    if (text.slice(i, i + 4) === 'null') {
      i += 4;
      return null;
    }
    error('Invalid token');
  }

  function parseNumber() {
    const start = i;
    if (text[i] === '-') i++;
    // integer part
    if (i >= n || text[i] < '0' || text[i] > '9') error('Invalid number');
    if (text[i] === '0') {
      i++;
      // leading zero must not be followed by another digit (e.g. "01")
      if (i < n && text[i] >= '0' && text[i] <= '9') error('Invalid number: leading zero');
    } else {
      while (i < n && text[i] >= '0' && text[i] <= '9') i++;
    }
    // fractional part
    if (i < n && text[i] === '.') {
      i++;
      if (i >= n || text[i] < '0' || text[i] > '9') error('Invalid number: missing fraction digits');
      while (i < n && text[i] >= '0' && text[i] <= '9') i++;
    }
    const numStr = text.slice(start, i);
    return Number(numStr);
  }

  // top-level
  skipWhitespace();
  if (i >= n) error('Empty input');
  const value = parseValue();
  skipWhitespace();
  if (i < n) error('Unexpected trailing characters');
  return value;
}

module.exports = { parseJSON };

File: problem1-json-parser/results/Python/A/main.py
class JSONParseError(Exception):
    pass


class _Parser:
    def __init__(self, text):
        self.text = text
        self.pos = 0
        self.n = len(text)

    def parse(self):
        self._skip_ws()
        value = self._parse_value()
        self._skip_ws()
        if self.pos != self.n:
            raise JSONParseError("trailing characters at %d" % self.pos)
        return value

    def _skip_ws(self):
        while self.pos < self.n and self.text[self.pos] in " \t\n\r":
            self.pos += 1

    def _peek(self):
        if self.pos < self.n:
            return self.text[self.pos]
        return None

    def _parse_value(self):
        c = self._peek()
        if c is None:
            raise JSONParseError("unexpected end of input")
        if c == '"':
            return self._parse_string()
        if c == '{':
            return self._parse_object()
        if c == '[':
            return self._parse_array()
        if c == 't' or c == 'f':
            return self._parse_bool()
        if c == 'n':
            return self._parse_null()
        if c == '-' or c.isdigit():
            return self._parse_number()
        raise JSONParseError("unexpected token %r at %d" % (c, self.pos))

    def _expect_literal(self, literal, result):
        if self.text[self.pos:self.pos + len(literal)] == literal:
            self.pos += len(literal)
            return result
        raise JSONParseError("invalid literal at %d" % self.pos)

    def _parse_null(self):
        return self._expect_literal("null", None)

    def _parse_bool(self):
        if self._peek() == 't':
            return self._expect_literal("true", True)
        return self._expect_literal("false", False)

    def _parse_string(self):
        # assumes current char is opening quote
        self.pos += 1
        chars = []
        while True:
            if self.pos >= self.n:
                raise JSONParseError("unterminated string")
            c = self.text[self.pos]
            if c == '"':
                self.pos += 1
                return "".join(chars)
            if c == '\\':
                self.pos += 1
                if self.pos >= self.n:
                    raise JSONParseError("unterminated escape")
                e = self.text[self.pos]
                if e == '"':
                    chars.append('"')
                elif e == '\\':
                    chars.append('\\')
                elif e == 'n':
                    chars.append('\n')
                else:
                    raise JSONParseError("invalid escape %r" % e)
                self.pos += 1
            else:
                chars.append(c)
                self.pos += 1

    def _parse_number(self):
        start = self.pos
        if self._peek() == '-':
            self.pos += 1
        # integer part
        if self.pos >= self.n or not self.text[self.pos].isdigit():
            raise JSONParseError("invalid number at %d" % start)
        if self.text[self.pos] == '0':
            self.pos += 1
            # leading zero must not be followed by more digits
            if self.pos < self.n and self.text[self.pos].isdigit():
                raise JSONParseError("invalid number (leading zero) at %d" % start)
        else:
            while self.pos < self.n and self.text[self.pos].isdigit():
                self.pos += 1
        is_float = False
        # fractional part
        if self.pos < self.n and self.text[self.pos] == '.':
            is_float = True
            self.pos += 1
            if self.pos >= self.n or not self.text[self.pos].isdigit():
                raise JSONParseError("invalid number (trailing dot) at %d" % start)
            while self.pos < self.n and self.text[self.pos].isdigit():
                self.pos += 1
        raw = self.text[start:self.pos]
        if is_float:
            return float(raw)
        return int(raw)

    def _parse_array(self):
        self.pos += 1  # consume [
        result = []
        self._skip_ws()
        if self._peek() == ']':
            self.pos += 1
            return result
        while True:
            self._skip_ws()
            result.append(self._parse_value())
            self._skip_ws()
            c = self._peek()
            if c == ',':
                self.pos += 1
                continue
            if c == ']':
                self.pos += 1
                return result
            raise JSONParseError("expected ',' or ']' at %d" % self.pos)

    def _parse_object(self):
        self.pos += 1  # consume {
        result = {}
        self._skip_ws()
        if self._peek() == '}':
            self.pos += 1
            return result
        while True:
            self._skip_ws()
            if self._peek() != '"':
                raise JSONParseError("expected string key at %d" % self.pos)
            key = self._parse_string()
            self._skip_ws()
            if self._peek() != ':':
                raise JSONParseError("expected ':' at %d" % self.pos)
            self.pos += 1
            self._skip_ws()
            value = self._parse_value()
            result[key] = value
            self._skip_ws()
            c = self._peek()
            if c == ',':
                self.pos += 1
                continue
            if c == '}':
                self.pos += 1
                return result
            raise JSONParseError("expected ',' or '}' at %d" % self.pos)


def parse_json(text):
    return _Parser(text).parse()
